### DISCUSSION TITLE
Offline Tokens & Offline Session Max Limited

### DISCUSSION THREAD

**Discussion by codyharris-h2o-ai** [keycloak/keycloak#26841] [Ideas]
Hello!

This is a continued discussion from the original issue at https://github.com/keycloak/keycloak/issues/26477.

## Background

We use offline tokens as credentials for our CLI.  Users log into the web interface, and navigate to a page, and get an offline refresh token, which they enter into their CLI during a setup process.  We also leverage these tokens in our Jenkins automation - tokens are configured into Jenkins as credentials, which is considered essentially read only.

## Offline Session Lifecycle

Currently, offline tokens may be used indefinitely provided they are used once every `$OFFLINE_SESSION_IDLE`.  We had wanted to enforce a global maximum lifetime for offline tokens, such that if unused, they expire within 7 days, but otherwise expire every 30 days.  We believed that setting  `Offline Session Max Limited` to enabled would enable this.

## Observations

When enabling `Offline Session Max Limited`, an `exp` (expiry) claim is added to the token, calculated by the current time + `$OFFLINE_SESSION_IDLE`.  This behaviour means that the token, regardless of usage, is only valid for `$OFFLINE_SESSION_IDLE` (rendering `$OFFLINE_SESSION_MAX` ineffective)

## Proposal for Consideration

* Add a toggle that enables a maximum lifespan for an offline session token (or maybe a radio button to pick from the 3 behaviours)
* Keep the existing behaviour that the token is valid if used once every `$OFFLINE_SESSION_IDLE`, it remains valid, and if not, it is considered expired
* Add an `exp` claim to the token that clearly marks when the token is no longer valid (based on `$OFFLINE_SESSION_MAX`)

**Comment by y-tabata** [comment]
Is it not acceptable to set both Offline Session Max and Offline Session Idle to 30 days? Is there any reason to insist on setting the expiration date to 7 days in the case of tokens are unused?
